torch/distributed/_composable/fully_shard.py

```diff
@@ -96,7 +99,7 @@ def fully_shard(
     _register_root_pre_forward_hook(state, module)  # prepend last
     for submodule in module.modules():
         if (
-            submodule not in state._ignored_modules
+            submodule.__class__.__name__ in state._fully_sharded_module_names
```

Comment:
My main concern is whether using `.__class__.name__` is too brittle. 

In my understanding, your PR uses `_fully_sharded_module_names` only for the `fully_shard` path to determine for which submodules you should call `_insert_module_state(submodule, state)`.

1. If we ignore parameters from one `nn.Linear`, then will we skip calling `_insert_module_state()` for _all_ `nn.Linear`? (because `"Linear"` will be in `_fully_sharded_module_names`)
2. If we ignore parameters from one `nn.Linear`, then will we skip calling `_insert_module_state()` for _all_ module classes called `Linear` (even if not from `nn`)? (because they all have `.__class__.__name__ == "Linear"`)

```
>>> lin1 = nn.Linear(3, 3)  # use built-in from nn
>>> class Linear(nn.Module):  # define own
>>>     def __init__(self):
>>>         super().__init__()
>>> lin2 = Linear()
>>> lin1.__class__.__name__
'Linear'
>>> lin2.__class__.__name__
'Linear'
```

The first question is a more important concern since it seems to affect correctness directly.

Reply:
That is a good point! I think the proper way is to use '_fully_sharded_module_to_handles' instead of '_fully_sharded_module_names' that is originally written for debugging (even though it is not ideal for debugging as well). 

One minor issue for using '_fully_sharded_module_to_handles' is that: the fully_sharded_module without parameters/handles will not have a fsdp state, it should be fine... as it does not have parameters to be managed by fsdp any way? I'm just worry about whether some edge cases I may miss?